com.unity.ml-agents/Runtime/Sensors/Reflection/ObservableAttribute.cs

```diff
@@ -0,0 +1,237 @@
+    /// </remarks>
+    /// <example>
+    /// This sample class will produce two observations, one for the m_Health field, and one
+    /// for the HealthPercent property.
+    /// <code>
+    /// using Unity.MLAgents;
+    /// using Unity.MLAgents.Sensors.Reflection;
+    ///
+    /// public class MyAgent : Agent
+    /// {
+    ///     [Observable]
+    ///     int m_Health;
+    ///
+    ///     [Observable]
+    ///     float HealthPercent
+    ///     {
+    ///         get => return 100.0f * m_Health / float(m_MaxHealth);
+    ///     }
+    /// }
+    /// </code>
+    /// </example>
+    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property)]
+    public class ObservableAttribute : Attribute
+    {
+        string m_Name;
+        int m_NumStackedObservations;
+
+        const BindingFlags k_BindingFlags = BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic;
+        static Dictionary<Type, int> s_TypeSizes = new Dictionary<Type, int>()
+        {
+            {typeof(int), 1},
+            {typeof(bool), 1},
+            {typeof(float), 1},
+            {typeof(Vector2), 2},
+            {typeof(Vector3), 3},
+            {typeof(Vector4), 4},
+            {typeof(Quaternion), 4},
+        };
+
+        /// <summary>
+        /// ObservableAttribute constructor.
+        /// </summary>
+        /// <param name="name">Optional override for the sensor name. Note that all sensors for an Agent
+        /// must have a unique name.</param>
+        /// <param name="numStackedObservations">Number of frames to concatenate observations from.</param>
+        public ObservableAttribute(string name = null, int numStackedObservations = 1)
+        {
+            m_Name = name;
+            m_NumStackedObservations = numStackedObservations;
+        }
+
+        internal static IEnumerable<(FieldInfo, ObservableAttribute)> GetObservableFields(object o, bool declaredOnly)
+        {
+            // TODO cache these (and properties) by type, so that we only have to reflect once.
+            var bindingFlags = k_BindingFlags | (declaredOnly ? BindingFlags.DeclaredOnly : 0);
+            var fields = o.GetType().GetFields(bindingFlags);
+            foreach (var field in fields)
+            {
+                var attr = (ObservableAttribute)GetCustomAttribute(field, typeof(ObservableAttribute));
+                if (attr != null)
+                {
+                    yield return (field, attr);
+                }
+            }
+        }
+
+        internal static IEnumerable<(PropertyInfo, ObservableAttribute)> GetObservableProperties(object o, bool declaredOnly)
+        {
+            var bindingFlags = k_BindingFlags | (declaredOnly ? BindingFlags.DeclaredOnly : 0);
+            var properties = o.GetType().GetProperties(bindingFlags);
+            foreach (var prop in properties)
+            {
+                if (!prop.CanRead)
+                {
+                    // Ignore write-only properties.
```

Comment:
Raise error ? Or warning ?

Reply:
Let me try to make this more consistent. I think the best behavior is to log warnings in the BehaviorParameters inspector (already have this for unsupported types) and ignore at runtime (currently write-only is ignored but unsupported throws).

Reply:
Fixed - now both write-only and unsupported data types produce warnings in BehaviorParameters inspector and are ignored at runtime.